pkg/sql/schemachanger/rel/database.go

```diff
@@ -103,6 +103,41 @@ func (t *Database) Insert(v interface{}) error {
 	return nil
 }
 
+// Delete deletes an entity.
+func (t *Database) Delete(v interface{}) error {
+	e, err := toEntity(t.schema, v)
+	if err != nil {
+		return err
+	}
+	if err := t.delete(e); err != nil {
+		return err
+	}
+	for _, v := range e.m {
+		_, isEntity := t.schema.entityTypeSchemas[reflect.TypeOf(v)]
+		_, alreadyDefined := t.entities[v]
+		if isEntity && alreadyDefined {
+			if err := t.Delete(v); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (t *Database) delete(e *entity) error {
+	self := e.getComparableValue(t.schema, Self)
+
+	delete(t.entities, self)
+	for i := range t.indexes {
+		idx := &t.indexes[i]
+		idx.tree.Delete(&containerItem{
+			entity:    e,
+			indexSpec: &idx.indexSpec,
+		})
+	}
+	return nil
```

Comment:
My first instict was that we should factor our a pattern here. Instead, though, I think we should see if this entity exists in the entities map. If it doesn't, just return nothing. If it does, iterate through its members and delete them rather than re-decomposing the entity.